mitiq/zne/inference.py

```diff
@@ -541,19 +558,27 @@ def run(
         # Get all noise-scaled circuits to run
         to_run = self._generate_circuits(qp, scale_noise, num_to_average)
 
-        # Get the list of keywords associated to each circuit in "to_run"
-        kwargs_list = self._get_keyword_args(num_to_average)
-
-        if Executor.is_batched_executor(executor):
-            if all([kwargs == {} for kwargs in kwargs_list]):
-                res = executor(to_run)
-            else:
-                res = executor(to_run, kwargs_list=kwargs_list)
-        else:
+        if observable is not None:
             res = [
-                executor(circ, **kwargs)  # type: ignore
-                for circ, kwargs in zip(to_run, kwargs_list)
+                observable.expectation(circuit, executor) for circuit in to_run
             ]
```

Comment:
In this way, in the case in which the `observable` argument is used, `kwargs_list` is ignored. 
If this is a temporary solution it's fine since `kwargs_list` is rarely used.

If instead we want to support it I guess we need to change `observable.expectation` to allow for *kwargs.
Alternatively,  things would be easy if we had an `observable.expectation_from_quantum_result` method which could be handy also in other contexts. Basically it would represent the concept of "trace of rho with O".

Anyway, if it's too complicated I would not block this PR.

Reply:
This gets a little subtle, see #952. It's worth discussing that issue separately from this so as to not settle on something suboptimal to get this merged - it's really about the behavior of `Executor.run` and not so much about ZNE. I added a TODO note with a link to this issue. Let me know what you think.